### PR TITLE
add cluster status collector

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards Node Pools. With this PR we replicate functionality of the `statusresource`, but for tenant clusters being managed upon `Cluster` CRs.

### CODE REVIEW

File: service/collector/cluster.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+func (c *Cluster) Collect(ch chan<- prometheus.Metric) error {
+	list, err := c.cmaClient.ClusterV1alpha1().Clusters(corev1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, cluster := range list.Items {
+		{
+			latest := key.ClusterCommonStatus(cluster).LatestCondition()
```
Comment (xh3b4sd) [inline service/collector/cluster.go]: This works now provider independent. More info below.

File: service/collector/set.go [hunk 2]
```diff
@@ -23,16 +25,6 @@ type Set struct {
...
 func NewSet(config SetConfig) (*Set, error) {
-	if config.CertSearcher == nil {
-		return nil, microerror.Maskf(invalidConfigError, "%T.CertSearcher must not be empty", config)
-	}
-	if config.G8sClient == nil {
-		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
-	}
-	if config.Logger == nil {
-		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
-	}
```
Comment (xh3b4sd) [inline service/collector/set.go]: This is not of any real use.

File: service/controller/clusterapi/v17/key/cluster_provider.go [hunk 3]
```diff
@@ -41,6 +41,30 @@ func g8sClusterSpecFromCMAClusterSpec(cmaSpec cmav1alpha1.ProviderSpec) (g8sv1al
 	return g8sSpec, nil
 }
 
+func g8sClusterCommonStatusFromCMAClusterStatus(cmaStatus *runtime.RawExtension) g8sv1alpha1.CommonClusterStatus {
+	type wrapper struct {
+		Cluster g8sv1alpha1.CommonClusterStatus `json:"cluster" yaml:"cluster"`
+	}
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v17/key/cluster_provider.go]: The implementation of the other methods here is provider specific which as we discussed are a bit off already. For the collector I wanted to have something more generic. Whatever provider status we unmarshal here, the wrapper should extract the common cluster status types that way.

File: service/service_test.go [hunk 4]
```diff
@@ -4,72 +4,8 @@ import (
...
 	"reflect"
 	"testing"
-
-	"github.com/giantswarm/micrologger/microloggertest"
-	"github.com/spf13/viper"
-
-	"github.com/giantswarm/cluster-operator/flag"
 )
 
-func Test_Service_New(t *testing.T) {
```
Comment (xh3b4sd) [inline service/service_test.go]: We dropped these tests in the other projects already as they are rather useless. Whoever things these are important, please scream.

**Comment by xh3b4sd** [comment]
Fishing for LGTM.

File: service/controller/clusterapi/v17/key/cluster_provider.go (hunk 3 above)
Comment (tuommaki) [reply service/controller/clusterapi/v17/key/cluster_provider.go]: This is actually a very clever way to deal with this. :+1:

Comment (kopiczko) [reply service/controller/clusterapi/v17/key/cluster_provider.go]: > Whatever provider status we unmarshal here, the wrapper should extract the common cluster status types that way.

I'd like to have that as a comment here.

Comment (xh3b4sd) [reply service/controller/clusterapi/v17/key/cluster_provider.go]: Added the comment.

**Pull request merged**